src/par_iter/internal.rs

```diff
@@ -9,133 +9,126 @@ use super::len::*;
 
 pub trait ProducerCallback<ITEM> {
     type Output;
-    fn callback<'p, P>(self, producer: P, shared: &'p P::Shared) -> Self::Output
-        where P: Producer<'p, Item=ITEM>;
+    fn callback<'p, P>(self, producer: P) -> Self::Output
+        where P: Producer<Item=ITEM>;
 }
 
 /// A producer which will produce a fixed number of items N. This is
 /// not queryable through the API; the consumer is expected to track
 /// it.
-pub trait Producer<'produce>: Send {
-    type Item;
-    type Shared: Sync + 'produce;
-
+pub trait Producer: IntoIterator + Send {
     /// Cost to produce `len` items, where `len` must be `N`.
-    fn cost(&mut self, shared: &Self::Shared, len: usize) -> f64;
+    fn cost(&mut self, len: usize) -> f64;
 
     /// Split into two producers; one produces items `0..index`, the
     /// other `index..N`. Index must be less than `N`.
-    unsafe fn split_at(self, index: usize) -> (Self, Self);
-
-    /// Unless a panic occurs, expects to be called *exactly N times*.
-    unsafe fn produce(&mut self, shared: &Self::Shared) -> Self::Item;
+    fn split_at(self, index: usize) -> (Self, Self);
 }
 
-/// A consumer which consumes items within the lifetime `'consume`.
-pub trait Consumer<'consume>: Send {
-    type Item: 'consume;
-    type Shared: Sync + 'consume;
-    type SeqState;
+/// A consumer which consumes items that are fed to it.
+pub trait Consumer<Item>: Send {
+    type Folder: Folder<Item, Result=Self::Result>;
+    type Reducer: Reducer<Self::Result>;
     type Result: Send;
 
     /// If it costs `producer_cost` to produce the items we will
     /// consume, returns cost adjusted to account for consuming them.
-    fn cost(&mut self, shared: &Self::Shared, producer_cost: f64) -> f64;
+    fn cost(&mut self, producer_cost: f64) -> f64;
 
     /// Divide the consumer into two consumers, one processing items
-    /// `0..index` and one processing items from `index..`.
-    unsafe fn split_at(self, shared: &Self::Shared, index: usize) -> (Self, Self);
+    /// `0..index` and one processing items from `index..`. Also
+    /// produces a reducer that can be used to reduce the results at
+    /// the end.
+    fn split_at(self, index: usize) -> (Self, Self, Self::Reducer);
+
+    /// Convert the consumer into a folder that can consume items
+    /// sequentially, eventually producing a final result.
+    fn fold(self) -> Self::Folder;
```

Comment:
`fold` is kind of a confusing name, given the type signature. How about constraining `Consumer` to also implement `Into<Self::Folder>`?